### PR TITLE
Protect (firewall) cluster using firewalld

### LINKED ISSUE #221: Firewalld (iptables) integration
We should add optional firewalld protection for infrastructures that does not have easily configurable firewall/network rules.

- https://www.firewalld.org/
- https://github.com/kubernetes/contrib/blob/master/ansible/roles/node/tasks/firewalld.yml

### PR DESCRIPTION
**Pull request by jakolehm**
```yaml
network:
  firewalld:
    enabled: true
    open_ports: # these are the defaults
    - port: '22'
      protocol: tcp
      roles:
      - "*"
    - port: '80'
      protocol: tcp
      roles:
      - worker
    - port: '443'
      protocol: tcp
      roles:
      - worker
    - port: '6443'
      protocol: tcp
      roles:
      - master
    - port: 30000-32767
      protocol: tcp
      roles:
      - "*"
    - port: 30000-32767
      protocol: udp
      roles:
      - "*"
    trusted_subnets: [] # array of strings
```

- [x] add cluster addresses to trusted zone
- [x] add 22/80/443/6443 to default zone
- [x] flag to enable/disable firewall (disabled by default)
- [x] node ports
- [x] custom ports/sources

Fixes #221

### CODE REVIEW

File: lib/pharos/scripts/configure-firewalld.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+add_port "80/tcp" "public"
+add_port "443/tcp" "public"
+add_port "30000-32767/tcp" "public"
+
+if [ "$ROLE" = "master" ]; then
+    add_port "6334/tcp" "public"
+fi
+
+for peer in $PEER_ADDRESSES ; do
+    add_source $peer "trusted"
```
Comment (kke) [inline lib/pharos/scripts/configure-firewalld.sh]:
```suggestion
    add_source "$peer" "trusted"
```

File: lib/pharos/scripts/configure-firewalld.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
+
+set -e
+
+add_port() {
+    local port=$1
+    local zone=$2
+    if ! firewall-cmd --query-port $port --zone $zone ; then
```
Comment (kke) [inline lib/pharos/scripts/configure-firewalld.sh]:
```suggestion
    if ! firewall-cmd --query-port "$port" --zone "$zone" ; then
```

**Comment by jakolehm** [comment]
@kke any ideas how to remove a host from whitelist (if host is removed from cluster.yml)?

File: lib/pharos/scripts/configure-firewalld.sh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        firewall-cmd --permanent --add-port $port --zone $zone
+    fi
+}
+
+add_source() {
+    local source=$1
+    local zone=$2
+    if ! firewall-cmd --query-source $source --zone $zone ; then
+        firewall-cmd --permanent --add-source $source --zone $zone
```
Comment (kke) [inline lib/pharos/scripts/configure-firewalld.sh]:
```suggestion
        firewall-cmd --permanent --add-source "$source" --zone "$zone"
```

**Comment by kke** [comment]
Maybe `iptables -F` but it sounds a bit risky

File: lib/pharos/scripts/configure-firewalld.sh [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    local zone=$2
+    if ! firewall-cmd --query-port "$port" --zone "$zone" ; then
+        firewall-cmd --permanent --add-port $port --zone $zone
+    fi
+}
+
+add_source() {
+    local source=$1
+    local zone=$2
+    if ! firewall-cmd --query-source $source --zone $zone ; then
```
Comment (kke) [inline lib/pharos/scripts/configure-firewalld.sh]:
```suggestion
    if ! firewall-cmd --query-source "$source" --zone "$zone" ; then
```

File: lib/pharos/scripts/configure-firewalld.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
+
+set -e
+
+add_port() {
+    local port=$1
+    local zone=$2
+    if ! firewall-cmd --query-port "$port" --zone "$zone" ; then
+        firewall-cmd --permanent --add-port $port --zone $zone
```
Comment (kke) [inline lib/pharos/scripts/configure-firewalld.sh]:
```suggestion
        firewall-cmd --permanent --add-port "$port" --zone "$zone"
```

**Comment by jakolehm** [comment]
> Maybe there is some kind of "flush all" option in firewalld that could be run in the beginning of the script

Did not find that kind of option... but maybe there is a command to remove all sources from a zone 🤔Ofc we could somehow loop all existing sources and remove them at the beginning?

**Comment by kke** [comment]
We could generate a list of removed nodes from previous configmap?

**Comment by jakolehm** [comment]
> We could generate a list of removed nodes from previous config?

@jnummelin protipped that generating firewalld conf file might be easier way to achieve this.

**Comment by jakolehm** [comment]
Now it generates service + ipset configs.

**Comment by jakolehm** [comment]
It seems that pod communication is busted if firewalld is first enabled and then disabled (it probably wipes too many iptables rules). How should we handle this? Error out if user tries to disable firewall?

**Comment by jakolehm** [comment]
@kke @jnummelin PTAL (again).

**Pull request merged**